### ISSUE TITLE
Create a master branch for AWS Distro Development

### ISSUE THREAD

**Issue by surajkota** [awslabs/kubeflow-manifests#76]
## Background

Current development process being followed for AWS-Kubeflow-distro-1.3.1 was to take a release branch from the manifest repository as the base (a.k.a upstream) and make AWS distribution specific changes on the same branch. This approach presented the fastest way to get development going considering the timeline, kfctl tool getting deprecated and vendors requiring to setup their own repository to host distribution specific manifests. 

## Problem

However, the above approach presents several issues:

1. Working on a new release:
    1. Every release requires a developer to create a copy of upstream KF release branch and copying/cherry-picking changes from the awslabs previous release branch (e.g. v1.3-branch) to new branch (e.g. v1.4-branch) which is cumbersome and error prone (e.g. #75)
    2. These release branches on upstream are not frozen because they are used for any patch releases. Let’s take an example for v1.4.0 and v1.4.1. Every patch release will require us to either:
        1. Create a new branch(e.g. v1.4.1-branch) from the tag/v1.4.1 from upstream and repeat the process in 1.a.
        2. Cherry pick commits related to patch release from upstream and create a PR to awslabs repo
        3. Rebase the awslabs/v1.4-branch from upstream/v1.4-branch and force push which is unacceptable
2. Maintenance issues unrelated to the code owned by the team
    1. Security vulnerabilities for test dependencies coming from upstream
    2. Modifying the root pkg READMEs on each release branch (e.g. [see README changes](https://github.com/awslabs/kubeflow-manifests/pull/42/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5))
3. AWS release version is coupled with Kubeflow version
    1. With this approach, we need to maintain 2 release versions one from Kubeflow and another from aws because we may have to do patch releases or feature enhancements e.g. aws-manifests-1.x-Kubeflow-1.3.1
4. Other issues:
    1. PR build and test infra(when we have one) will need to be constantly be updated to point to different branches instead of one master branch. Similarly Github issue templates need to be ported
    2. Default branch on the repository will keep changing
    3. Might loose the history related to the PRs between releases if changes are copied instead of cherry-picking

## Suggested Approach

Majority of the overlays do not change unless there are major changes in the upstream component. Create a base with AWS overlays that does not change between versions and work only on the delta per release.

### Development - Build and Test

Create a master branch to host only AWS specific overlays, tests and docs. e.g.

```
ubuntu@ip-172-31-0-119:~/kubeflow-manifests$ tree -L 3
.
├── README.md
├── awsconfig
│   ├── apps
│   │   ├── jupyter
│   │   ├── katib
│   │   └── pipelines
│   ├── common
│   │   ├── alb-controller
│   │   ├── istio-envoy-filter
│   │   └── istio-ingress
│   └── infra_configs
├── docs
│   ├── cognito
│   ├── rds
│   ├── s3
│   └── storage
├── tests
│   ├── canary
│   ├── e2e
│   └── unit

```

(**Customer experience)** This will introduce one additional step for users after cloning the awslabs repo.

```
export KUBEFLOW_RELEASE_VERSION=v1.4.1
git clone --branch ${KUBEFLOW_RELEASE_VERSION} https://github.com/kubeflow/manifests.git upstream
```

Once user has cloned a tagged branch from upstream manifests(kubeflow/manifests repo), they will continue with setting up aws resources or running kustomize build and apply just like today.

For this, we need to isolate the aws specific changes in form of overlays to a separate directory.

* I have already completed a [POC](https://github.com/awslabs/kubeflow-manifests/pulls?q=is%3Apr+author%3Asurajkota+v1.4) for jupyter, cognito, rds and s3 in the v1.4-branch as a POC. There should be no changes for EFS and FSx since the changes are self contained. The only piece left is secrets-manager integration which is also doable

### Releases

New releases (major and patch from upstream): Cut a release once a Kubeflow release has been validated with updated README. 

Patch for an old release(depends on support policy) where latest overlays are incompatible: Use the existing release tag branch to do a patch release or backport features for older releases.

* Follow up: Will this base become outdated between several releases?
    * The intent of this development strategy is not to support multiple versions using the latest build. Making changes to a previous release will take the same amount of time i.e. go to release branch and make the changes.

### How does this solve the problems stated earlier?

We take out the problems related to PR build/test, copying changes over from old release branches and modifying parent READMEs etc by creating this base package with only AWS manifests instead of using a Kubeflow released branch as base. The contributors only needs to work on the delta per release. Problems related to security vulnerabilities because of unrelated packages are handled by only cloning the upstream during build time.

Patch releases just become a small PR with changing a version in a config file and re-running the tests using the PR build infra.

The problem for release version is not completely solved since a it also depends on the upstream but we are able to cut a new release for this package independently which talks about the compatible version of Kubeflow (TBD - see below)

### Next steps/TODOs

* How to document/convery the supported upstream version to users?
* What is the release versioning for this package?
* Use the [main branch](https://github.com/awslabs/kubeflow-manifests/tree/main) to get started with this

Please comment your thoughts/feedback

**Comment by AlexandreBrown** [comment]
Hello @surajkota , this approach is very reasonable.  
On our side, we came up to a solution that is pretty similar to isolate the files we needed to change for pre-install (files that are applied during the install) and post-install (yaml that must be applied after kubeflow is installed). This way when a new release comes, we can easily diff our pre and post-install files with the new version and overwrite the new version files with ours if everything is fine (this can be done in minutes).  
<img src="https://user-images.githubusercontent.com/26939775/150593562-05a82188-de17-413f-aae7-264edebe9318.png" height="400">  
Therefore I think your proposal makes total sense.
  
We are on board with your proposal as is and we're eager to install kubeflow 1.4.x when this proposal will be in place.
